### ISSUE TITLE
UserError.Abort can cause MissingFormatArgumentException

### ISSUE THREAD

**Issue by DeanWookey** [oracle/graal#2836]
**Describe the issue**
Building a native image with a generated reflection configuration file as well as jni-config.json, proxy-config.json, and resource-config.json (as described here: https://github.com/oracle/graal/blob/master/substratevm/Configuration.md) may cause an error which causes the UserError.Abort method to fail, resulting in an unhelpful error message.

**Steps to reproduce the issue**
I'm unable to reproduce this since I don't know what's causing the error in my case due to the unhelpful MissingFormatArgumentException error message.

**Describe GraalVM and your environment:**
 - GraalVM version **graalvm-ce-java11-windows-amd64-20.2.0**
 - JDK major version: **11**
 - OS: **Windows 10**
 - Architecture: **AMD64**

```
Fatal error:java.util.MissingFormatArgumentException: Format specifier '%5c'
	at java.base/java.util.Formatter.format(Formatter.java:2672)
	at java.base/java.util.Formatter.format(Formatter.java:2609)
	at java.base/java.lang.String.format(String.java:2897)
	at com.oracle.svm.core.util.UserError.abort(UserError.java:68)
	at com.oracle.svm.hosted.config.ConfigurationParserUtils.doParseAndRegister(ConfigurationParserUtils.java:138)
	at com.oracle.svm.hosted.config.ConfigurationParserUtils.lambda$parseAndRegisterConfigurations$3(ConfigurationParserUtils.java:119)
	at java.base/java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:212)
	at com.oracle.svm.hosted.config.ConfigurationParserUtils$1.tryAdvance(ConfigurationParserUtils.java:108)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:312)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.IntPipeline.reduce(IntPipeline.java:491)
	at java.base/java.util.stream.IntPipeline.sum(IntPipeline.java:449)
	at com.oracle.svm.hosted.config.ConfigurationParserUtils.parseAndRegisterConfigurations(ConfigurationParserUtils.java:125)
	at com.oracle.svm.reflect.hosted.ReflectionFeature.duringSetup(ReflectionFeature.java:69)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$setupNativeImage$12(NativeImageGenerator.java:876)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:876)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:553)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:468)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Comment by oubidar-Abderrahim** [comment]
Could you please follow these steps to build the native-image

1. make sure that the application compiles and executes correctly. (sometimes the error is in the code, we just miss it).
2. run the application again this time using the command `java -agentlib:native-image-agent=config-output-dir=path_to_the_folder_in_which_you_want_to_put_the_generated_config_files -jar your_application.jar` (please make sure to test as much use cases as possible in your application while using the agent, this way it guarantees to have higher coverage of your application and generate more complete config files).
3. generate a native-image, either by providing the flags in the command or using `native-image.properties` file. make sure it has the following flags: `--no-fallback` `-H:+ReportExceptionStackTraces` and `-H:+StackTrace` (usually native-image can detect config files if put in META-INF/native-image. But to be sure, it's better for diagnosis to specify the path each time) so add also `-H:ReflectionConfigurationFiles=path-to-file/reflect-config.json` `-H:ResourceConfigurationFiles=path-to-file/resource-config.json` `-H:DynamicProxyConfigurationResources=path-to-file/proxy-config.json` and `-H:JNIConfigurationResources=path-to-file/jni-config.json` (you may wanna add `-H:+JNI` flag if you're using JNI).
4. If the image generation succeeded you can run your app and test if it works correctly.

Please let me know if you still get the error with the native-image after this.

**Comment by DeanWookey** [comment]
I think I have managed to get past that point (I have another error now, which I'm working through).

The bug report is not to do with the native image failing, but to do with the fault reporting itself crashing.

If you look at the code relevant to this line: com.oracle.svm.hosted.config.ConfigurationParserUtils.doParseAndRegister(ConfigurationParserUtils.java:138)

```java
    private static void doParseAndRegister(ConfigurationParser parser, Reader reader, String featureName, Object location, HostedOptionKey<String[]> option) {
        try {
            parser.parseAndRegister(reader);
        } catch (IOException | JSONParserException e) {
            String errorMessage = e.getMessage();
            if (errorMessage == null || errorMessage.isEmpty()) {
                errorMessage = e.toString();
            }
            throw UserError.abort("Error parsing " + featureName + " configuration in " + location + ":\n" + errorMessage +
                            "\nVerify that the configuration matches the schema described in the " +
                            SubstrateOptionsParser.commandArgument(PrintFlags, "+") + " output for option " + option.getName() + ".");
        }
    }
```
It's trying to throw a meaningful error message, however in that method it crashes because one of the variables in that line contains '%5c' apparently. It should probably be changed to:

```java
UserError.abort("Error parsing %s configuration in %s:\n%s\nVerify that the configuration matches the schema described in the %s output for option %s.",featureName,location,errorMessage,SubstrateOptionsParser.commandArgument(PrintFlags, "+"),option.getName());
```

I believe my specific problem had something to do with the arguments you mention (-H:ReflectionConfigurationFiles=path-to-file/reflect-config.json etc.) I'm building for javaFX using gluon's maven plugin which was adding all those args in. I followed their docs on how to add mine so that their plugin could combine them. That seems to have solved my issue.

**Comment by oubidar-Abderrahim** [comment]
@cstancu can you take a look at this.

**Comment by jaikiran** [comment]
I've opened https://github.com/oracle/graal/pull/2838 with a potential fix

**Comment by peter-hofer** [comment]
Fixed by 725d8ab3e820143146dc778b3aaaeb2825d5c7e2.
